### ISSUE TITLE
OneToOne is not cascaded correctly

### ISSUE THREAD

**Issue by jsprw** [mikro-orm/mikro-orm#210]
**Describe the bug**
If you create a new object with a OneToOne relation and persist that object, the items are persisted but the relation id's are still empty.

For example:

```typescript
const newEnt1 = wrap(new Enitity1()).assign({
        prop1: req.body.prop1,
        ent2: {
                prop: req.body.prop,
        }
});
```
This creates a row for ent2 in the database but it doesn't fill the foreign keys (ent2.ent1_id is NULL) (doesn't work if ent1 is owner of the relation nor if ent2 is. I expected something like the collections, that it fills the id's as well.

Am I doing something wrong or is this not supported?

**Comment by B4nan** [comment]
Can I see those entities as well?

**Comment by B4nan** [comment]
Might be just a bug with `assign`, have you tried doing it manually?

**Comment by jsprw** [comment]
> Might be just a bug with `assign`, have you tried doing it manually?

Manually assigning like 

```typescript 
ent1.ent2 = {
    prop: req.body.prop,
}
```
doesn't work either.

> Can I see those entities as well?

```
//entity1
@OneToOne(() => Entity2, e => e.ent1)
    ent2: Entity2;

//entity2
@OneToOne(() => Entity1)
    ent1: Entity1;
```

**Comment by B4nan** [comment]
The problem here is that you are assigning to the inverse side only (`Entity1`). For 1:1 you need to work with the relation from the owning side. For collections (1:m/m:n) there is propagation working as it is possible thanks to the collection wrapper, but for 1:1 it is not possible. Although it could be done when using `assign` helper like you do, that is probably possible... 

Cascading will most probably work ok for updating existing entities, but when creating, you will need to handle it manually.

**Comment by jsprw** [comment]
I do only have one owning side. I have switched it the other way around and now I get this error:
`Unhandled Promise rejection: Cannot read property 'isInitialized' of undefined ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'isInitialized' of undefined` at UnitOfWork.cascadeReference

**Comment by B4nan** [comment]
> I do only have one owning side

There is always only one owning side - the one that stores the reference in database. 

Please add code snippet so I can reproduce the last issue too. If you work with owning side, it does work, I did try so you might be doing something wrong. 

I will try to support this when using assign or em.create, should be possible. But the persisting itself is broken a bit as I noted above.

**Comment by jsprw** [comment]
> Please add code snippet so I can reproduce the last issue too. If you work with owning side, it does work, I did try so you might be doing something wrong.

It is the same code as above but with this switched:

```
//entity1
@OneToOne(() => Entity2)
    ent2: Entity2;

//entity2
@OneToOne(() => Entity1, e => e.ent2)
    ent1: Entity1;
```

Then it throws the error.

**Comment by B4nan** [comment]
> Manually assigning like
>
> ent1.ent2 = {
>     prop: req.body.prop,
> }

This is wrong, you need to assign entity instance, not JS object (entity.assign() does that for you). Then it should work from the owning side at least. 

    ent1.ent2 = new Entity2({
      prop: req.body.prop,
    })

**Comment by jsprw** [comment]
No, this code is used, not ent1.ent2:

```
const newEnt1 = wrap(new Enitity1()).assign({
        prop1: req.body.prop1,
        ent2: {
                prop: req.body.prop,
        }
});
```

**Comment by jsprw** [comment]
Since when does this code work btw?

```
new Entity2({
  prop: req.body.prop,
})
```

**Comment by jsprw** [comment]
Or it throws this error:

```
Unhandled Promise rejection: Cannot read property 'type' of null ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'type' of null
    at UnitOfWork.commitChangeSet
```

**Comment by B4nan** [comment]
Yes, and I suggested to do it manually, without `entity.assign()`, then you posted the code I quoted (it was from your message). By manually I meant that code I posted:

```typescript
const ent1 = new Entity1(); // owning side
ent1.ent2 = new Entity2();
await em.persistAndFlush(ent1);
```

To make it clear:

```typescript
//entity1
@OneToOne(() => Entity2)
ent2: Entity2; // owning side

//entity2
@OneToOne(() => Entity1, e => e.ent2)
ent1: Entity1; // inverse side
```

> Since when does this code work btw?

That was just for illustration, depends on how you define your constructor of course.

**Comment by B4nan** [comment]
> Or it throws this error:

This is the issue I was talking about when you have it referenced bidirectionally.

**Comment by jsprw** [comment]
> This is the issue I was talking about when you have it referenced bidirectionally.

In this piece of code I didn't reference it bidirectionally.

> Yes, and I suggested to do it manually, without entity.assign(), then you posted the code I quoted (it was from your message). By manually I meant that code I posted:

This code works for me indeed. Sorry that I didn't understand. Would be nice to have it included in assign indeed. :)

**Comment by B4nan** [comment]
Should be fixed in `3.0.0-alpha.32`. 

https://github.com/mikro-orm/mikro-orm/commit/a1320baf3fece75de44199798acfbf65366e82a4#diff-fe75d4396c979dec70ee161d766bd6deR233-R247

**Comment by jsprw** [comment]
It seems to be working now. On deeply nested properties, it sill throws a maximum stack size exceeded error.

**Comment by B4nan** [comment]
What code is causing it? Does specifying the type argument manually helps?

**Comment by jsprw** [comment]
I'm specifying the type argument everywhere. It happens when the entity creates another nested entity automatically because it has new Entity3() as default value for a property.

Entity 3 has on its part another property that links back to the new cascaded Entity2. And while logging the names that are handled, it seems that it keeps circulating between Entity2 -> Entity3 -> Entity2 -> Entity3, etc.

**Comment by B4nan** [comment]
Add the code so I can replicate it pls.

**Comment by jsprw** [comment]
Hopefully below is enough information:

```
const newEnt1 = wrap(new Enitity1()).assign({
        ent2: {
                prop: req.body.prop,
        }
});

...
.persistAndFlush(newEnt1);

//entity1
@OneToOne(() => Entity2, e => e.e1_id)
    ent2: Entity2;

//entity2
@OneToOne(() => Entity3)
    ent3: Entity3 = new Entity3();

@Property({type:"string"})
   prop: string;

@OneToOne(() => Entity1)
    e1_id: Entity1;

//entity3
@OneToOne(() => Entity2, e => e.ent3)
    parent: Entity2;
```

**Comment by B4nan** [comment]
I can't replicate this, works as expected for me, all entities are properly cascaded and stored. Where do you get this error from? Please post a stack trace. Or ideally prepare reproduction repository. 

Generally would be nice if you follow the bug report issue template, I am asking for that information because it will help me a lot when understanding what happens. I am not even sure what driver you are using.

**Comment by B4nan** [comment]
If the recursion issue you are talking about is the same as in the reproduction repo from #222 than it should be fixed in `3.0.0-alpha.33`

**Comment by jsprw** [comment]
Yes. Thank you very much again! Looks like it was the same issue and does not occur anymore.
